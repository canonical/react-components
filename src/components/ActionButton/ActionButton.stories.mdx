import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";

import ActionButton from "./ActionButton";

<Meta
  title="ActionButton"
  component={ActionButton}
  argTypes={{
    children: {
      control: {
        type: "text",
      },
    },
  }}
/>

### ActionButton

This is a not an existing Vanilla component. It can be used to display submitting states for forms or other actions.

### Props

ActionButton accepts the props from [Button](?path=/docs/button--base) in addition to the following:

<ArgsTable of={ActionButton} />

### Default

<Canvas>
  <Story
    name="Default"
    args={{
      appearance: "positive",
      children: "Click me!"
    }}
  >
    {ActionButton.bind({})}
  </Story>
</Canvas>

### Loading

<Canvas>
  <Story
    name="Loading"
    args={{
      appearance: "positive",
      loading: true,
      disabled: true,
      children: "Click me!"
    }}
  >
    {ActionButton.bind({})}
  </Story>
</Canvas>
