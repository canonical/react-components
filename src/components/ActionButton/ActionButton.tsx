import classNames from "classnames";
import React, { useEffect, useRef, useState } from "react";
import type { ButtonHTMLAttributes, ReactNode } from "react";

import type { ButtonProps } from "../Button";
import Icon from "../Icon";

import type { ClassName, PropsWithSpread } from "types";

export const LOADER_MIN_DURATION = 400; // minimium duration (ms) loader displays
export const SUCCESS_DURATION = 2000; // duration (ms) success tick is displayed

export enum Label {
  WAITING = "Waiting for action to complete",
  SUCCESS = "Action completed",
}

export type Props = PropsWithSpread<
  {
    /**
     * The appearance of the button.
     */
    appearance?: ButtonProps["appearance"];
    /**
     * The content of the button.
     */
    children?: ReactNode;
    /**
     * Optional class(es) to pass to the button element.
     */
    className?: ClassName;
    /**
     * Whether the button should be disabled.
     */
    disabled?: boolean;
    /**
     * Whether the button should display inline.
     */
    inline?: boolean;
    /**
     * Whether the button should be in the loading state.
     */
    loading?: boolean;
    /**
     * Whether the button should be in the success state.
     */
    success?: boolean;
  },
  ButtonHTMLAttributes<HTMLButtonElement>
>;

/**
 * This is a not an existing Vanilla component. It can be used to display submitting states for forms or other actions.
 *
 * ActionButton accepts the props from
 * [Button](?path=/docs/components-button--docs) in addition to those in the
 * props table:
 */
const ActionButton = ({
  appearance,
  children,
  className,
  disabled = false,
  inline = false,
  loading = false,
  success = false,
  ...buttonProps
}: Props): JSX.Element => {
  const [height, setHeight] = useState<number | null>();
  const [width, setWidth] = useState<number | null>();
  const [showLoader, setShowLoader] = useState(false);
  const [showSuccess, setShowSuccess] = useState(false);
  const ref = useRef<HTMLButtonElement>(null);

  // Set up loader timer
  useEffect(() => {
    let loaderTimeout: number;

    if (loading) {
      // Explicitly set button dimensions
      if (ref.current && !!ref.current.getBoundingClientRect()) {
        setHeight(ref.current.getBoundingClientRect().height);
        setWidth(ref.current.getBoundingClientRect().width);
      }
      setShowLoader(true);
    }

    if (!loading && showLoader) {
      loaderTimeout = window.setTimeout(() => {
        setShowLoader(false);

        if (success) {
          setShowSuccess(true);
        }
      }, LOADER_MIN_DURATION);
    }

    if (!loading && !showLoader) {
      setHeight(null);
      setWidth(null);
    }

    return () => window.clearTimeout(loaderTimeout);
  }, [loading, showLoader, success]);

  // Set up success timer
  useEffect(() => {
    let successTimeout: number;

    if (showSuccess) {
      successTimeout = window.setTimeout(() => {
        setHeight(null);
        setWidth(null);
        setShowSuccess(false);
      }, SUCCESS_DURATION);
    }

    return () => window.clearTimeout(successTimeout);
  }, [showSuccess]);

  const buttonClasses = classNames(
    className,
    "p-action-button",
    appearance ? `p-button--${appearance}` : "p-button",
    {
      "is-processing": showLoader || showSuccess,
      "is-disabled": disabled,
      "is-inline": inline,
    }
  );
  const showIcon = showLoader || showSuccess;
  const icon = (showLoader && "spinner") || (showSuccess && "success") || null;
  const iconLight = appearance === "positive" || appearance === "negative";

  // This component uses the base button element instead of the Button component
  // as the button requires a ref and Button would have to be updated to use
  // forwardRef which is not currently supported by components that use
  // typescript generics.
  return (
    <button
      className={buttonClasses}
      disabled={disabled}
      ref={ref}
      style={
        height && width
          ? {
              height: `${height}px`,
              width: `${width}px`,
            }
          : undefined
      }
      {...buttonProps}
    >
      {showIcon ? (
        <Icon
          aria-label={showLoader ? Label.WAITING : Label.SUCCESS}
          className={showLoader ? "u-animation--spin" : null}
          light={iconLight}
          name={icon}
        />
      ) : (
        children
      )}
    </button>
  );
};

export default ActionButton;
