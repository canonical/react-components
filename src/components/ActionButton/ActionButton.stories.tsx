import { Meta, StoryObj } from "@storybook/react";

import ActionButton from "./ActionButton";

const meta: Meta<typeof ActionButton> = {
  component: ActionButton,
  tags: ["autodocs"],

  argTypes: {
    children: {
      control: {
        type: "text",
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof ActionButton>;

export const Default: Story = {
  name: "Default",

  args: {
    appearance: "positive",
    children: "Click me!",
  },
};

export const Loading: Story = {
  name: "Loading",

  args: {
    appearance: "positive",
    loading: true,
    disabled: true,
    children: "Click me!",
  },
};
