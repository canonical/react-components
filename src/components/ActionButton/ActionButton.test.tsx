import React from "react";
import { render, screen } from "@testing-library/react";

import ActionButton, {
  Label,
  LOADER_MIN_DURATION,
  SUCCESS_DURATION,
} from "./ActionButton";
import { act } from "react-dom/test-utils";

describe("ActionButton", () => {
  jest.useFakeTimers("modern");

  it("matches loading snapshot", () => {
    render(<ActionButton loading>Click me</ActionButton>);
    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("matches success snapshot", () => {
    const { rerender } = render(<ActionButton loading>Click me</ActionButton>);
    // Move on to the success state.
    rerender(<ActionButton success>Click me</ActionButton>);
    act(() => {
      jest.advanceTimersByTime(LOADER_MIN_DURATION + 1);
    });
    expect(screen.getByRole("button")).toMatchSnapshot();
  });

  it("shows a loader icon for the correct amount of time", () => {
    const { rerender } = render(<ActionButton loading>Click me</ActionButton>);
    // Check for the accessible element.
    expect(screen.getByLabelText(Label.WAITING)).toBeInTheDocument();
    const icon = ".p-icon--spinner";
    // Check for the visual element.
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelector(icon)).toBeInTheDocument();
    // Move on to the success state.
    rerender(<ActionButton success>Click me</ActionButton>);
    act(() => {
      jest.advanceTimersByTime(LOADER_MIN_DURATION);
    });
    expect(screen.queryByLabelText(Label.WAITING)).not.toBeInTheDocument();
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelector(icon)).not.toBeInTheDocument();
  });

  it("shows a success icon for the correct amount of time", () => {
    const { rerender } = render(<ActionButton loading>Click me</ActionButton>);
    // Move on to the success state.
    rerender(<ActionButton success>Click me</ActionButton>);
    act(() => {
      jest.advanceTimersByTime(LOADER_MIN_DURATION + 1);
    });
    // Check for the accessible element.
    expect(screen.getByLabelText(Label.SUCCESS)).toBeInTheDocument();
    const icon = ".p-icon--success";
    // Check for the visual element.
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelector(icon)).toBeInTheDocument();
    act(() => {
      jest.advanceTimersByTime(SUCCESS_DURATION + 1);
    });
    expect(screen.queryByLabelText(Label.SUCCESS)).not.toBeInTheDocument();
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelector(icon)).not.toBeInTheDocument();
  });
});
