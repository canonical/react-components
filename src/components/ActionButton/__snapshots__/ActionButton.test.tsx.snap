// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButton matches loading snapshot 1`] = `
<ActionButton
  loading={true}
>
  <button
    className="p-action-button p-button is-processing"
    disabled={false}
  >
    <Icon
      className="u-animation--spin"
      light={false}
      name="spinner"
    >
      <i
        className="u-animation--spin p-icon--spinner"
      />
    </Icon>
  </button>
</ActionButton>
`;

exports[`ActionButton matches success snapshot 1`] = `
<ActionButton
  loading={false}
  success={true}
>
  <button
    className="p-action-button p-button is-processing"
    disabled={false}
  >
    <Icon
      className={null}
      light={false}
      name="success"
    >
      <i
        className="p-icon--success"
      />
    </Icon>
  </button>
</ActionButton>
`;
